### PR TITLE
RequestHeader authentication: add UID to recognized request headers

### LINKED ISSUE #93699: UserInfo.UID is not passed by client-go
It looks UserInfo.UID is newer than other fields and while #49677 added it SubjectAccessReviewSpec for webhook authorization, it's still not being sent by client-go in headers. Also there is no flag in APIServer to process those headers. As a result all authorizers needing that, may not work correctly for SubjectAccessReviews made by kube-aggregator extensions.

**What happened**:

Our webhook authorizer, authorizes users using UserInfo.UID. It's missing in SubjectAccessReviews created by kube-aggregator extensions (our case was state metrics), resulting authorizer to not work.

**What you expected to happen**:

I except All fields that is filled by authenticator be present for authorizer and I think it's valid to have an authorizer to use UID.

**How to reproduce it (as minimally and precisely as possible)**:

Make authentication webhook that fills SubjectAccessReviewSpec.UID. You will see there UID is missing from SubjectAccessReviewSpecs made by a kube-aggregator extension.

**Anything else we need to know?**:

UserInfo.UID is not passed to client-go round-tripper [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/transport/round_trippers.go#L118) and [here](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/transport/round_trippers.go#L101) and impersonate config do nothing for it. There is also no api-server flag to get it from headers.

### PR DESCRIPTION
**Pull request by stlaz**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
This PR adds the ability to set a UID header in order to be recognized in the RequestHeader authentication. It pushes the new header configuration into the `kube-system/extension-apiserver-authentication` CM, and makes the client-go AuthProxyRoundTripper add user UID to the requests it handles.

#### Which issue(s) this PR fixes:

Fixes #93699

#### Does this PR introduce a user-facing change?

```release-note
kube-apiserver: a new `--requestheader-uid-headers` flag allows configuring request header authentication to obtain the authenticating user's UID from the specified headers. The suggested value for the new option is `X-Remote-Uid`. When specified, the `kube-system/extension-apiserver-authentication` configmap will include the value in its `.data[requestheader-uid-headers]` field.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by enj** [comment]
@stlaz I haven't looked at the diff yet but remember that you need to add an integration test that shows this works correctly with aggregated API servers.  I would recommend using a service account to call an aggregated API and maybe intercept the authorization checks made by the aggregated API server to see what identity it saw.

**Comment by stlaz** [comment]
Yes, an integration test is currently missing. I suppose I could write some kind of an aggregated API server similar to how the admission webhooks are being tested:
https://github.com/kubernetes/kubernetes/blob/aef1832fbf9e584b5fefa5d90717d585a3340643/test/integration/apiserver/admissionwebhook/admission_test.go#L473

**Comment by stlaz** [comment]
The front-proxy config test now lives separately from the other, rather big aggregated apiserver test.

**Comment by valaparthvi** [comment]
Hi @stlaz. I am from the Bug triage shadow team for K8s release 1.27. I am checking in to make sure this PR is on track for 1.27 release.

Please note that Code Freeze begins on March 15, post this date, only critical and release-blocker issues/PRs will be accepted.

File: staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go [hunk 1]
```diff
@@ -19,9 +19,10 @@ package headerrequest
 import (
 	"context"
 	"encoding/json"
-	"k8s.io/apimachinery/pkg/api/equality"
 	"testing"
 
+	"k8s.io/apimachinery/pkg/api/equality"
+
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go:25]: Delete empty line

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go [hunk 2]
```diff
@@ -97,6 +101,9 @@ func (s *RequestHeaderAuthenticationOptions) AddFlags(fs *pflag.FlagSet) {
 	fs.StringSliceVar(&s.UsernameHeaders, "requestheader-username-headers", s.UsernameHeaders, ""+
 		"List of request headers to inspect for usernames. X-Remote-User is common.")
 
+	fs.StringSliceVar(&s.UIDHeaders, "requestheader-uid-headers", s.UIDHeaders, ""+
+		"List of request headers to inspect for UIDs. X-Remote-UID is suggested.")
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:105]: Update all `UID` in header keys to be `Uid` to follow how Go canonicalizes these headers.

```suggestion
		"List of request headers to inspect for UIDs. X-Remote-Uid is suggested.")
```

File: test/integration/examples/apiserver_test.go [hunk 3]
```diff
@@ -214,9 +232,118 @@ func TestAggregatedAPIServer(t *testing.T) {
...
+	if len(saToken) == 0 {
+		t.Fatal("empty SA token in token request response")
+	}
+
+	// wrap the authz round tripper to catch the request for our SA SAR to the KAS
+	wardleOptions.RecommendedOptions.Authorization.WithCustomRoundTripper(
+		// adding a round tripper wrapper to test default RequestHeader configuration
+		transport.WrapperFunc(func(rt http.RoundTripper) http.RoundTripper {
+			return roundTripperFunc(func(req *http.Request) (*http.Response, error) {
+				gotUser, _ := genericapirequest.UserFrom(req.Context())
```
Comment (enj) [inline test/integration/examples/apiserver_test.go:110]: Check the `ok` here.

File: test/integration/examples/apiserver_test.go [hunk 4]
```diff
@@ -56,167 +62,122 @@ import (
...
+				if gotUser.GetName() == "system:aggregator" {
+					return rt.RoundTrip(req)
+				}
+
+				if len(gotUser.GetUID()) == 0 {
+					t.Errorf("expected UID to be non-empty for user %q", gotUser.GetName())
+				}
+				if got, expected := gotUser.GetUID(), expectedSAUserInfo.GetUID(); expected != got {
+					t.Errorf("expected UID: %q, got: %q", expected, got)
+				}
```
Comment (enj) [inline test/integration/examples/apiserver_test.go:122]: I ran this test against `master` and it fails as expected:

```
    apiserver_test.go:118: expected UID to be non-empty for user "system:serviceaccount:integration-test-front-proxy-config:wardle-client-sa"
    apiserver_test.go:121: expected UID: "792c2633-ec56-4b7d-a25c-5417d8824251", got: ""
```

**Review by enj — CHANGES_REQUESTED** [review]
Not done yet but initial feedback.

**Comment by enj** [comment]
/release-note-edit

```release-note
A new option is being added to services that are capable of using the request header authentication - `requestheader-uid-headers`.

This new option allows you to configure which headers should be used in order to recognize the authenticating user's UID.

The suggested value for the new option is `X-Remote-Uid`.

The `kube-system/extension-apiserver-authentication` config map, that is commonly used for cluster trust discovery, will reflect this new option in its `.Data` under the `requestheader-uid-headers` key.
```

File: test/integration/examples/apiserver_test.go [hunk 5]
```diff
@@ -56,167 +62,122 @@ import (
...
+		// adding a round tripper wrapper to test default RequestHeader configuration
+		transport.WrapperFunc(func(rt http.RoundTripper) http.RoundTripper {
+			return roundTripperFunc(func(req *http.Request) (*http.Response, error) {
+				gotUser, _ := genericapirequest.UserFrom(req.Context())
+
+				// this is likely the KAS checking the OpenAPI endpoints
+				if gotUser.GetName() == "system:aggregator" {
+					return rt.RoundTrip(req)
+				}
+
```
Comment (enj) [inline test/integration/examples/apiserver_test.go:116]: Increment an atomic counter here and assert that it is equal to `1` at the end of the test (so we know these assertions are actually called).

**Comment by stlaz** [comment]
/test pull-kubernetes-integration
Odd, the TestFrontProxy test appears to have failed, but then again it seems that there was some kind of a network issue?
I can't reproduce the failure locally, though.

**Comment by valaparthvi** [comment]
Hello,
It's me again, Parthvi from the Bug Triage Release team 1.28 👋. Since this PR is a spill over from the last milestone, I wanted to check in early this time and make sure it's still on track for 1.28. Please let me know if there's any help you require from my :)

Thanks.

**Comment by valaparthvi** [comment]
Hello again,

With Code Freeze coming in, on [01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496), please make sure this PR goes in by then. We are currently in Week 7, and the code free will be here by Week 9.
Do let me know if I can be of any help :)

**Comment by valaparthvi** [comment]
Hello @stlaz,
I am checking in on behalf of the Bug Triage release team. We are only a week away from Code Freeze ([01:00 UTC Wednesday 19th July 2023 / 18:00 PDT Tuesday 18th July 2023](https://everytimezone.com/s/72ee8496)). Please ensure your PR goes in by then if you are targeting 1.28 release. 

Please let me know if I can be of any assistance :)

**Comment by MaryamTavakkoli** [comment]
Hello! 
I am checking in on behalf of the Bug Triage release team.
This PR has not been updated for a long time, so I'd like to check what's the status. The code freeze is starting (01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023) (about 2 weeks from now). Please ensure your PR goes in by then if you are targeting 1.29 release. (Also, please add the priority label)

Please let us know if we can be of any assistance and help :)

**Comment by MaryamTavakkoli** [comment]
Hello!
Bug triage for the 1.29 release cycle is here! I want to check the status and remind you that the code freeze is starting [01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/24b27b03) (next week), as we want to ensure that each PR has a chance to be merged.
Are we still targeting the 1.29 release?

File: staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go:1]: Do we have a test case for the upgrade scenario where the existing CM has no UID config, but the API server now does have said config, meaning it should update the config map.

**Review by enj — COMMENTED** [review]
Comments from today's review meeting:

```
request header authn

proxy --(RH)--> KAS

this proxy is telling KAS who the user is
EKS -> request signing

KAS --(RH)--> aggregated API server

the proxy is KAS, and it is telling the aggregated server what the user is

bug:

david forgot that UID is a thing

username
UID
groups
extra fields (metadata)

for example, authz knows about UID

proxy --(RH)--> server
mTLS
you can constrain to a single CN
```

It is unclear how to make a change like this safe on upgrade if we want to update the defaults.

**Comment by enj** [comment]
> It is unclear how to make a change like this safe on upgrade if we want to update the defaults.

Maybe we just have to leave KAS with no default value here, but can update the defaults for anything "behind" KAS (i.e. aggregated API servers) because the only actor that can assert that mTLS connection is KAS, and we can rely on KAS deleting the UID headers from any incoming request before it forwards the request onwards.

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go [hunk 6]
```diff
@@ -116,6 +120,9 @@ func (s *RequestHeaderAuthenticationOptions) AddFlags(fs *pflag.FlagSet) {
 	fs.StringSliceVar(&s.UsernameHeaders, "requestheader-username-headers", s.UsernameHeaders, ""+
 		"List of request headers to inspect for usernames. X-Remote-User is common.")
 
+	fs.StringSliceVar(&s.UIDHeaders, "requestheader-uid-headers", s.UIDHeaders, ""+
+		"List of request headers to inspect for UIDs. X-Remote-Uid is suggested.")
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:124]: Per triage meeting review from today:

This flag needs to not be wired up in the first release that contains this feature as then it would be possible for old KAS to proxy a request to a new aggregated API server without dropping the UID header.  Everything else is safe to do in this PR.  To be able to test this code path in integration tests, we can use an approach similar to `SetServiceResolverForTests` to let us set the `UIDHeaders` field.  A future PR in a later release would simply drop that global in place of the actual flag wiring.

I believe it is safe to update `NewDelegatingAuthenticationOptions` with the `X-Remote-Uid` as a default because no matter what that relies on the front proxy dropping headers that the client should not set.  And that code path isn't really used because the vast majority of cases should be using the config map, which will only get updated once KAS is updated and the `--requestheader-uid-headers` CLI flag is set.  @liggitt any thoughts on the upgrade story here?

Once we let people configure this flag on KAS, we should require that it contains at least `X-Remote-Uid` so that API aggregation will work correctly (we can be stricter here because the config is new).

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:124]: Just got here, need to think about the skew implications.

There was just a lengthy discussion about what skew between kube-apiserver and aggregated servers are allowed / recommended / supported (https://github.com/kubernetes/website/pull/46109, https://github.com/kubernetes/kubernetes/issues/124533, https://github.com/kubernetes/kubernetes/issues/124655, api-machinery meeting https://www.youtube.com/watch?v=0TXm-DGcK1k&t=31m). We don't really dictate whether aggregated API servers' use of k8s.io/apiserver library can / should be newer than kube-apiserver today. cc @deads2k as well who might have thoughts.

I agree our defaults (in kube-apiserver and k8s.io/apiserver library) should not lead to aggregated API servers trusting headers _before_ kube-apiserver sets / guards / protects them.

I think the following defaults work:
* kube-apiserver 1.31+ does not default the uid CLI flag (just like the other requestheader flags)
* kube-apiserver 1.31+ (as kube-aggregator) unconditionally clears the standard uid header when proxying (just like it clears the other standard headers)
* kube-apiserver 1.31+ (as kube-aggregator) unconditionally sets the standard uid header when proxying (just like it sets the other standard headers)
* k8s.io/apiserver library v0.31.0+ wires the uid CLI flag
* k8s.io/apiserver library v0.31.0+ NewDelegatingAuthenticationOptions() does not default the uid header; this keeps the CLI flags safe against much older kube-apiservers which aren't protecting the header; in practice, most aggregated API servers pull their auth header config from the configmap published by kube-apiserver rather than the CLI flags, so if/when their kube-apiserver publishes uid header config, they'll start using it

In the future (kube-apiserver 1.32+) we can consider making kube-apiserver also include the standard username/groups/extra/uid headers the aggregator uses (which are not configurable) in the configmap it publishes containing auth config, so that aggregation works even if kube-apiserver only set non-standard requestheader flag options

File: staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go
Comment (stlaz) [reply staging/src/k8s.io/apiserver/pkg/authentication/request/headerrequest/requestheader_controller_test.go:1]: I added a unit test that verifies that the controller should be capable of this step.

Looking at the controller, we might still face trouble for the upgrade period where kube-apiserver would exist in different versions, it doesn't seem to be guarded by leader election. That might lead to update hotloops.

The same issue would occur today if somebody wanted to configure additional headers, although that's probably not very typical action, compared to cluster upgrades.

I suspect we may want to address this. Not sure if adding leader election to KAS controllers is an OK practice?

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go (hunk 6 above)
Comment (stlaz) [reply staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:127]: I removed the defaulting from the `DelegatingAuthenticationOptions()` which I think was the only thing missing to get us to the desired state as described by Jordan.

We'll probably still have to think how to update the cluster trust CM - https://github.com/kubernetes/kubernetes/pull/115834#discussion_r1611467973.

File: test/integration/examples/apiserver_test.go [hunk 7]
```diff
@@ -655,6 +606,224 @@ func TestAggregatedAPIServerRejectRedirectResponse(t *testing.T) {
 	}
 }
 
+func prepareAggregatedWardleAPIServer(ctx context.Context, t *testing.T, namespace string, requestHeadersUIDOverride []string) (*kastesting.TestServer, *sampleserver.WardleServerOptions, int) {
```
Comment (enj) [inline test/integration/examples/apiserver_test.go:609]: Lets split the test refactor out into its own PR so we can merge that first and make this change easier to review.

File: cmd/kube-apiserver/app/testing/testserver.go [hunk 8]
```diff
@@ -87,7 +87,8 @@ type TestServerInstanceOptions struct {
 	// Useful for testing server configurations expected to prevent /healthz from completing.
 	SkipHealthzCheck bool
 	// Enable cert-auth for the kube-apiserver
-	EnableCertAuth bool
+	EnableCertAuth                  bool
+	RequestHeaderUIDHeadersOverride []string
```
Comment (enj) [inline cmd/kube-apiserver/app/testing/testserver.go:91]: Instead of wiring a new override config, lets merge the default requester header options with whatever is specified via flags (with the flags having precedence).

File: test/integration/examples/apiserver_test.go [hunk 9]
```diff
@@ -225,164 +232,143 @@ func TestAPIServiceWaitOnStart(t *testing.T) {
...
+func TestFrontProxyConfig(t *testing.T) {
+	const testNamespace = "integration-test-front-proxy-config"
+
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
 	t.Cleanup(cancel)
 
-	// makes the kube-apiserver very responsive.  it's normally a minute
-	dynamiccertificates.FileRefreshDuration = 1 * time.Second
+	// start up the KAS and prepare the options for the wardle API server
+	testKAS, wardleOptions, wardlePort := prepareAggregatedWardleAPIServer(ctx, t, testNamespace, []string{"x-remote-uid"})
```
Comment (enj) [inline test/integration/examples/apiserver_test.go:250]: Also add a test case where the UID header field is not set and thus the SA token UID is not sent to the aggregated API server.  In the future, this test will fail when we handle the defaulting logic in a way to separate KAS front proxy config vs. aggregated API server front proxy config

File: test/integration/examples/apiserver_test.go [hunk 10]
```diff
@@ -225,164 +232,143 @@ func TestAPIServiceWaitOnStart(t *testing.T) {
...
-func TestAggregatedAPIServer(t *testing.T) {
+type roundTripperFunc func(*http.Request) (*http.Response, error)
+
+func (f roundTripperFunc) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req)
+}
+
+// TestFrontProxyConfig tests that the RequestHeader configuration is consumed
+// correctly by the aggregated API servers.
+func TestFrontProxyConfig(t *testing.T) {
```
Comment (enj) [inline test/integration/examples/apiserver_test.go:243]: We also need a test to show that KAS itself can consume the UID header config (i.e. it should not consume UIDs unless the new flag is set).  I think the `certreload` integration tests may have most of the logic that you need.

**Comment by enj** [comment]
> The latest commit removes the `DelegatedAuthenticationOptions()` defaulting for the UID headers.

Let's include an inline comment that describes why we didn't update that function.

File: test/integration/examples/apiserver_test.go (hunk 7 above)
Comment (stlaz) [reply test/integration/examples/apiserver_test.go:609]: opened https://github.com/kubernetes/kubernetes/pull/126270 for that

File: test/integration/examples/apiserver_test.go (hunk 10 above)
Comment (stlaz) [reply test/integration/examples/apiserver_test.go]: added 2 integration tests for cases where remote UID headers are set and when they are not

**Comment by stlaz** [comment]
/test pull-kubernetes-unit
looks like somebody broke some fuzzing tests which are now flaking

**Comment by stlaz** [comment]
/test pull-kubernetes-unit
these appear as unrelated flakes

File: staging/src/k8s.io/client-go/transport/round_trippers_test.go [hunk 11]
```diff
@@ -349,7 +353,7 @@ func TestAuthProxyRoundTripper(t *testing.T) {
 	} {
 		rt := &testRoundTripper{}
 		req := &http.Request{}
-		NewAuthProxyRoundTripper(tc.username, tc.groups, tc.extra, rt).RoundTrip(req)
+		NewAuthProxyRoundTripper(tc.username, tc.uid, tc.groups, tc.extra, rt).RoundTrip(req)
```
Comment (enj) [inline staging/src/k8s.io/client-go/transport/round_trippers_test.go:356]: Linter is mad that the returned error is ignored.

File: test/integration/auth/requestheader_test.go [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+
+	caPrivateKey, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	caBytes, err := x509.CreateCertificate(rand.Reader, ca, ca, &caPrivateKey.PublicKey, caPrivateKey)
```
Comment (enj) [inline test/integration/auth/requestheader_test.go:151]: grep for `x509.CreateCertificate` in the codebase - we have multiple helpers in `client-go` and in `test/` for helping with these types of operations.

File: test/integration/auth/requestheader_test.go [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+				SetHeader("X-Remote-Uid", "test-uid").
+				SetHeader("X-Remote-User", "testuser").
+				SetHeader("X-Remote-Groups", "group1", "group2").
+				Do(tCtx).
+				Into(selfInfo)
+			if err != nil {
+				t.Fatalf("failed to retrieve self-info: %v", err)
+			}
+
+			if selfUID := selfInfo.Status.UserInfo.UID; (len(selfUID) != 0) != tt.setUID {
```
Comment (enj) [inline test/integration/auth/requestheader_test.go:126]: nit this is a bit hard to read:

```suggestion
			if hasUID := len(selfInfo.Status.UserInfo.UID) != 0; hasUID != tt.setUID {
```

File: test/integration/auth/requestheader_test.go [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+	tCtx := ktesting.Init(t)
+
+	frontProxyCA, frontProxyClient, frontProxyKey, err := newTestCAWithClient(
+		pkix.Name{
+			CommonName: "test-front-proxy-ca",
+		},
+		big.NewInt(43),
+		pkix.Name{
+			CommonName:   "test-aggregated-apiserver",
+			Organization: []string{"system:masters"},
```
Comment (enj) [inline test/integration/auth/requestheader_test.go:51]: We don't need this right?

```suggestion
```

File: test/integration/auth/requestheader_test.go [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+		},
+		big.NewInt(86),
+	)
+	if err != nil {
+		t.Error(err)
+		return
```
Comment (enj) [inline test/integration/auth/requestheader_test.go:57]:
```suggestion
		t.Fatal(err)
```

**Review by enj — COMMENTED** [review]
LGTM, minor comments.

File: staging/src/k8s.io/client-go/transport/round_trippers_test.go (hunk 11 above)
Comment (stlaz) [reply staging/src/k8s.io/client-go/transport/round_trippers_test.go]: added an explicit error ignore

File: test/integration/auth/requestheader_test.go (hunk 12 above)
Comment (stlaz) [reply test/integration/auth/requestheader_test.go]: I used some functions for generating the CA and marshalling the pkey. There is nothing for client certs in the lib, unfortunately, so that part has to stay.

Comment (enj) [reply test/integration/auth/requestheader_test.go]: Why can't we use `k8s.io/kubernetes/test/utils.NewSignedCert`?

File: test/integration/auth/requestheader_test.go (hunk 13 above)
Comment (enj) [reply test/integration/auth/requestheader_test.go:124]: (this nit is outstanding but I cannot tell if you meant to ignore it or forgot about it)

File: test/integration/auth/requestheader_test.go (hunk 12 above)
Comment (stlaz) [reply test/integration/auth/requestheader_test.go]: I must have missed that one as I was focusing mostly on the client-go library once I found the functions there. I'll use that one.

Would be nice to have a library for these things, this pki_utils code appears to be also copied to kubeadm.

**Comment by enj** [comment]
/lgtm
/approve
/assign liggitt 
For approval

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go [hunk 16]
```diff
@@ -80,6 +84,9 @@ func (s *RequestHeaderAuthenticationOptions) Validate() []error {
 	if len(s.UsernameHeaders) > 0 && !caseInsensitiveHas(s.UsernameHeaders, "X-Remote-User") {
 		klog.Warningf("--requestheader-username-headers is set without specifying the standard X-Remote-User header - API aggregation will not work")
 	}
+	if len(s.UIDHeaders) > 0 && !caseInsensitiveHas(s.UIDHeaders, "X-Remote-Uid") {
+		klog.Warningf("--requestheader-uid-headers is set without specifying the standard X-Remote-Uid header - API aggregation will not work")
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:88]: since this is a new header / flag, should we be more strict here and make this an error to start?

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go [hunk 17]
```diff
@@ -233,7 +244,12 @@ func NewDelegatingAuthenticationOptions() *DelegatingAuthenticationOptions {
 		CacheTTL:   10 * time.Second,
 		ClientCert: ClientCertAuthenticationOptions{},
 		RequestHeader: RequestHeaderAuthenticationOptions{
-			UsernameHeaders:     []string{"x-remote-user"},
+			UsernameHeaders: []string{"x-remote-user"},
+			// we specifically don't default UID headers as these were introduced
+			// later and we don't want 3rd parties to be trusting the default headers
+			// before we can safely say that all KAS instances know they should
+			// remove them from an incoming request in its WithAuthentication handler
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:251]: note the specific release KAS handles this (1.32+) and / or link to the issue tracking the promotion to default (https://github.com/kubernetes/kubernetes/issues/126821)

File: staging/src/k8s.io/client-go/transport/round_trippers.go [hunk 18]
```diff
@@ -131,6 +135,7 @@ func SetAuthProxyHeaders(req *http.Request, username string, groups []string, ex
 	}
 
 	req.Header.Set("X-Remote-User", username)
+	req.Header.Set("X-Remote-Uid", uid)
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/transport/round_trippers.go:138]: only if `len(uid) > 0`?

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go (hunk 16 above)
Comment (stlaz) [reply staging/src/k8s.io/apiserver/pkg/server/options/authentication.go:88]: Let's do it, that way we can always switch back to a warning in case somebody's able to find a plausible use case that doesn't use the standard header.

**Comment by stlaz** [comment]
implemented the suggestions + addressed the `CredentialIDKey` constant move from the serviceaccount module to user in the integration tests.

**Comment by sftim** [comment]
Changelog suggestion
```diff
-kube-apiserver: a new `--requestheader-uid-headers` flag allows configuring request header authentication to obtain the authenticating user's UID from the specified headers. The suggested value for the new option is `X-Remote-Uid`. When specified, the `kube-system/extension-apiserver-authentication` configmap will include the value in its `.data[requestheader-uid-headers]` field.
+kube-apiserver: Added a new command line argument, `--requestheader-uid-headers`. Setting this allows configuring request header authentication to obtain the authenticating user's UID from the specified headers. The suggested value for the new option is `Remote-Uid`. When specified, the `kube-system/extension-apiserver-authentication` configmap will include the value in its `.data[requestheader-uid-headers]` field.
```

I changed `X-Remote-Uid` to `Remote-Uid` as this matches what the IETF recommend about HTTP header names - IIRC at least.

File: pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go [hunk 19]
```diff
@@ -259,6 +262,10 @@ func getConfigMapDataFor(authenticationInfo ClusterAuthenticationInfo) (map[stri
 		if err != nil {
 			return nil, err
 		}
+		data["requestheader-uid-headers"], err = jsonSerializeStringSlice(authenticationInfo.RequestHeaderUIDHeaders.Value())
```
Comment (liggitt) [inline pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go:265]: this may be causing dueling / ~hotloop configmap writes when 1.31 and 1.32 API servers interact, see https://github.com/kubernetes/kubernetes/issues/128986#issuecomment-2516536495

Comment (dims) [reply pkg/controlplane/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go:265]: 😢 😭
